[Inductor] Fix combo kernel heuristics for persistent_reduction

karthickai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #168050
* #169721
* __->__ #169509
* #168939

This PR fixes combo kernel `size_hints` selection for multiple persistent_reduction kernels with different rnumel values. now correctly selects the kernel with maximum rnumel (e.g., `r0_=32` instead of `r0_=16`).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

karthickai committed:
[WIP][Inductor] Fix combo kernel heuristics for persistent_reduction

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Fix combo kernel heuristics for persistent_reduction"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Fix combo kernel heuristics for persistent_reduction"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Fix combo kernel heuristics for persistent_reduction"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Fix combo kernel heuristics for persistent_reduction"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[Inductor] Fix combo kernel heuristics for persistent_reduction"


This PR fixes combo kernel `size_hints` selection for multiple persistent_reduction kernels with different rnumel values. now correctly selects the kernel with maximum rnumel (e.g., `r0_=32` instead of `r0_=16`).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo mlazos

[ghstack-poisoned]

karthickai commented on torch/_inductor/codegen/triton_combo_kernel.py:
```diff
@@ -547,7 +547,14 @@ def select_combo_heuristics(
         else:
-            return heuristics_list[0], size_hints_list[0], self.sub_kernels[0]
+            # find persistent_reduction with maximum rnumel
+            i, _ = max(
```
combo kernels share a single xblock constexpr parameter for all branches. The autotuner generates configs based on size_hints that must work for all sub-kernels. On A10G, configs generated for size_hints={'r0_': 4} are optimized for small R0_BLOCK and exceed the register limit, causing "No valid triton configs" error (https://github.com/pytorch/pytorch/actions/runs/19627156075/job/56198492241). Using max rnumel ensures the autotuner generates configs with sufficient register budget to fit all sub-kernels within hardware limits.

eellison replied:
> single xblock constexpr parameter for all branches

Is this fundamental , though?

karthickai replied:
Yeah it is fundamental, I try to say that why we selected size_hint max rnumel for persistent reduction

karthickai committed:
Update on "[Inductor] Fix combo kernel heuristics for persistent_reduction"


This PR fixes combo kernel `size_hints` selection for multiple persistent_reduction kernels with different rnumel values. now correctly selects the kernel with maximum rnumel (e.g., `r0_=32` instead of `r0_=16`).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[Inductor] Fix combo kernel heuristics for persistent_reduction"


This PR fixes combo kernel `size_hints` selection for multiple persistent_reduction kernels with different rnumel values. now correctly selects the kernel with maximum rnumel (e.g., `r0_=32` instead of `r0_=16`).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo mlazos

[ghstack-poisoned]
